### ISSUE TITLE
building: incompatible types error in compute/cast/utf8_to.rs

### ISSUE THREAD

**Issue by fdncred** [pola-rs/polars#14134]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://crates.io/crates/polars) of Polars.

### Reproducible example

Just building the crate, these lines produce the error.
```rust
    let buffers = if uses_buffer {
        Arc::from([arr.values().clone()])
    } else {
        Arc::from([])
    };
```

### Log output

```shell
error[E0308]: `if` and `else` have incompatible types
   --> C:\Users\dschroeder\.cargo\registry\src\index.crates.io-6f17d22bba15001f\polars-arrow-0.37.0\src\compute\cast\utf8_to.rs:101:9
    |
98  |       let buffers = if uses_buffer {
    |  ___________________-
99  | |         Arc::from([arr.values().clone()])
    | |         --------------------------------- expected because of this
100 | |     } else {
101 | |         Arc::from([])
    | |         ^^^^^^^^^^^^^ expected an array with a fixed size of 1 element, found one with 0 elements
102 | |     };
    | |_____- `if` and `else` have incompatible types
    |
    = note: expected struct `Arc<[buffer::immutable::Buffer<u8>; 1]>`
               found struct `Arc<[_; 0]>`
```

### Issue description

I'm trying to update nushell to the latest polars 0.37.0 crates and cannot build because of this error.

### Expected behavior

clean build without warnings or errors.

### Installed versions

<details>

polars-io = { version = "0.37", features = ["avro"], optional = true }
polars-arrow = { version = "0.37", optional = true }
polars-ops = { version = "0.37", optional = true }
polars-plan = { version = "0.37", optional = true }
polars-utils = { version = "0.37", optional = true }

[dependencies.polars]
features = [
	"arg_where",
	"checked_arithmetic",
	"concat_str",
	"cross_join",
	"csv",
	"cum_agg",
	"default",
	"dtype-categorical",
	"dtype-datetime",
	"dtype-struct",
	"dtype-i8",
	"dtype-i16",
	"dtype-u8",
	"dtype-u16",
	"dynamic_group_by",
	"ipc",
	"is_in",
	"json",
	"lazy",
	"object",
	"parquet",
	"random",
	"rolling_window",
	"rows",
	"serde",
	"serde-lazy",
	"strings",
	"to_dummies",
]
optional = true
version = "0.37"

</details>

**Comment by ritchie46** [comment]
That's strange, what is your compiler version?

**Comment by fdncred** [comment]
We have a rust-toolchain.toml file that forces the compiler to this version currently. We try to stay 2 versions behind rust's latest so distros can keep up to date easier.
```toml
[toolchain]
profile = "default"
channel = "1.73.0"
```

**Comment by BGR360** [comment]
I have a minimal repro below. Looks like the problem only occurs when a local variable is introduced.

`arc-slice.rs`:

```rust
use std::sync::Arc;

fn takes_arc_slice(_: Arc<[u8]>) {}

fn works() {
    takes_arc_slice(if true {
        Arc::from([255])
    } else {
        Arc::from([])
    });
}

fn doesnt_work() {
    let buffers = if true {
        Arc::from([255])
    } else {
        Arc::from([])
    };
    takes_arc_slice(buffers);
}

fn main() {
    works();
    doesnt_work();
}

```

It fails to compile on 1.73.0 with the same error. It compiles successfully on 1.74.0 and stable (1.75.0). I'll file an issue on the Rust repo so they can at least put a regression test in there.

```
$ rustc +1.73.0 arc-slice.rs
error[E0308]: `if` and `else` have incompatible types
  --> arc-slice.rs:17:9
   |
14 |       let buffers = if true {
   |  ___________________-
15 | |         Arc::from([255])
   | |         ---------------- expected because of this
16 | |     } else {
17 | |         Arc::from([])
   | |         ^^^^^^^^^^^^^ expected an array with a fixed size of 1 element, found one with 0 elements
18 | |     };
   | |_____- `if` and `else` have incompatible types
   |
   = note: expected struct `Arc<[{integer}; 1]>`
              found struct `Arc<[_; 0]>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
$ rustc +1.74.0 arc-slice.rs
$ rustc +stable arc-slice.rs
$
```

**Comment by fdncred** [comment]
Thanks for the research @BGR360! Since a new rust release will be out in about a week (Feb-8th I think), and the compile works in 1.74, we can just wait to bump nushell to the next version of polars until then. I'm happy to have this issue closed if the powers that be wish it so. Much love to the team and contributors here ❤️!

**Comment by jmcnamara** [comment]
+1

I also came across this issue with `rustc` 1.72.0 when trying to upgrade the Polars version used by `rust_xlsxwriter`.

**Comment by LCrossman** [comment]
I encountered this error when running rust-matplotlib on Macbook M2 under Rosetta2 with rustc v 1.71.  However, there was no error when running the same library under arm64 in rustc v 1.78. 
I use anaconda package manager to switch between environments for the M2 to run as arm64 or through Rosetta2 run as x86_64.
I then updated rustc from 1.71 to 1.77 in x86_64 environment and the error was solved and now no polars arrow error in either environment.
